libmamba/src/core/env_lockfile.cpp

```diff
@@ -0,0 +1,214 @@
+// Copyright (c) 2022, QuantStack and Mamba Contributors
+//
+// Distributed under the terms of the BSD 3-Clause License.
+//
+// The full license is in the file LICENSE, distributed with this software.
+
+#include "mamba/core/env_lockfile.hpp"
+
+#include <yaml-cpp/yaml.h>
+#include <spdlog/fmt/fmt.h>
+
+#include "mamba/core/output.hpp"
+
+#include "mamba/core/match_spec.hpp"
+
+namespace mamba
+{
+    namespace env_lockfile_v1
+    {
+        using Package = EnvironmentLockFile::Package;
+
+        tl::expected<Package, mamba_error> read_package_info(const YAML::Node& package_node)
+        {
+            Package package{
+                /* .info = */ mamba::PackageInfo{ package_node["name"].as<std::string>() },
+                /* .is_optional = */
+                [&]
+                {
+                    if (const auto& optional_node = package_node["optional"])
+                        return optional_node.as<bool>();
+                    return false;
+                }(),
+                /* .category = */ package_node["category"].as<std::string>(),
+                /* .manager = */ package_node["manager"].as<std::string>(),
+                /* .platform = */ package_node["platform"].as<std::string>(),
+            };
+
+            package.info.version = package_node["version"].as<std::string>();
```

Comment:
Wouldn't this throw a YAML Exception if no version is available? Just wondering if we should catch the exception and transform to an expected?

Reply:
So far for any Yamlcpp exception, I assumed it's fine to catch it in `read_environment_lockfile` (below) instead of every functions, but I could catch them earlier and return expected indeed. In `read_environment_lockfile` the exception is already converted to an expected, though it's not visible from here so no problem if you prefer local catching instead.